test/modellib/ElementwiseModel.cpp

```diff
@@ -0,0 +1,205 @@
+
+#include "mlir/IR/BuiltinOps.h"
+
+#include "include/OnnxMlirRuntime.h"
+#include "src/Compiler/CompilerUtils.hpp"
+#include "src/Dialect/ONNX/ONNXOps.hpp"
+#include "src/Runtime/OMTensorHelper.hpp"
+#include "test/modellib/ModelLib.hpp"
+
+#include <functional>
+
+using namespace mlir;
+
+namespace onnx_mlir {
+namespace test {
+
+// =============================================================================
+// Elementwise without broadcast
+// Support following ops:
+//    ONNXAddOp
+//    ONNXHardSigmoidOp
+
+const float alphaVal = 2.0;
+const float betaVal = 0.5;
+
+Elementwise2DLibBuilder::Elementwise2DLibBuilder(const std::string &modelName,
+    const std::string &onnxOpName, const int inputNum, const int I, const int J)
+    : ModelLibBuilder(modelName), onnxOpName(onnxOpName), inputNum(inputNum),
+      I(I), J(J) {}
+
+bool Elementwise2DLibBuilder::build() {
+  llvm::SmallVector<int64_t, 4> shape = {I, J};
+  Type elementType = builder.getF32Type();
+  auto aType = RankedTensorType::get(shape, elementType);
+  auto bType = RankedTensorType::get(shape, elementType);
+  auto yType = RankedTensorType::get(shape, elementType);
+
+  llvm::SmallVector<Type, 2> inputsType{aType};
+  if (inputNum > 1)
+    inputsType.emplace_back(bType);
+  llvm::SmallVector<Type, 1> outputsType{yType};
+
+  func::FuncOp funcOp = createEmptyTestFunction(inputsType, outputsType);
+  Block &entryBlock = funcOp.getBody().front();
+
+  // Create operation.
+  llvm::SmallVector<Value, 1> results;
+  if (inputNum == 1) {
+    // Unary operations.
+    auto aVal = entryBlock.getArgument(0);
+    if (onnxOpName.compare("ONNXHardSigmoidOp") == 0) {
+      // Hard Sigmoid.
+      FloatAttr alpha = builder.getFloatAttr(elementType, alphaVal);
+      FloatAttr beta = builder.getFloatAttr(elementType, betaVal);
+      auto op =
+          builder.create<ONNXHardSigmoidOp>(loc, yType, aVal, alpha, beta);
+      results.emplace_back(op.getResult());
+    } else
+      llvm_unreachable("unsupported unary elementwise op");
+
+  } else if (inputNum == 2) {
+    // Binary operations.
+    auto aVal = entryBlock.getArgument(0);
+    auto bVal = entryBlock.getArgument(1);
+    if (onnxOpName.compare("ONNXAddOp") == 0) {
+      // Add.
+      assert(inputNum == 2 && "bad input number");
+      auto op = builder.create<ONNXAddOp>(loc, yType, aVal, bVal);
+      results.emplace_back(op.getResult());
+    } else
+      llvm_unreachable("unsupported binary elementwise op");
+  } else
+    llvm_unreachable("support only unary and binary op");
+
+  // Create function.
+  builder.create<func::ReturnOp>(loc, results);
+  module.push_back(funcOp);
+  createEntryPoint(funcOp);
+  return true;
+}
+
+bool Elementwise2DLibBuilder::prepareInputs(
+    float dataRangeLB, float dataRangeUB) {
+  // Create dynamic list (to be freed before end of function).
+  OMTensor **list = (OMTensor **)malloc(inputNum * sizeof(OMTensor *));
```

Comment:
PR #2078  tried to replace this `malloc` by `OMTensor* list[inputNum];`, recommend to use it here.

Reply:
I see. So the current code looks fine as you free them properly.

Reply:
Now use stack alloc to size 2, and may only use 1 when inputNum==1. Simpler code.